pkg/customizations/firstboot/firstboot.go

```diff
@@ -0,0 +1,163 @@
+	// https://aap.example.com/api/controller/v2/job_templates/9/callback/ or
+	// https://aap.example.com/api/v2/job_templates/9/callback/ depending on the
+	// AAP version.
+	JobTemplateURL string
+
+	// The host config key. Required if type is set to "aap".
+	HostConfigKey string
+}
+
+// FirstbootOption is a union of all supported firstboot options.
+type FirstbootOption interface {
+	isFirstbootOption()
+}
+
+func (CustomFirstbootOptions) isFirstbootOption()    {}
+func (SatelliteFirstbootOptions) isFirstbootOption() {}
+func (AAPFirstbootOptions) isFirstbootOption()       {}
+
+var ErrFirstbootAlreadySet = errors.New("firstboot customization already set")
+
+// FirstbootOptionsFromBP converts a blueprint FirstbootCustomization to
+// FirstbootOptions. Validation is done in the blueprint package, so this function
+// assumes the input is valid, however, JSON unmarshalling errors are possible.
+func FirstbootOptionsFromBP(bpFirstboot blueprint.FirstbootCustomization) (*FirstbootOptions, error) {
+	fo := &FirstbootOptions{}
+	var satDone, aapDone bool
+
+	for _, fbsc := range bpFirstboot.Scripts {
+		cust, sat, aap, err := fbsc.SelectUnion()
+		if err != nil {
+			return nil, err
+		}
+
+		if cust != nil {
+			opts := CustomFirstbootOptions{
+				FirstbootCommonOptions: FirstbootCommonOptions{
+					Name:          cust.Name,
+					IgnoreFailure: cust.IgnoreFailure,
+				},
+				Contents: cust.Contents,
+			}
+			fo.Scripts = append(fo.Scripts, opts)
+		}
+
+		if sat != nil {
+			opts := SatelliteFirstbootOptions{
+				FirstbootCommonOptions: FirstbootCommonOptions{
+					Name:          sat.Name,
+					IgnoreFailure: sat.IgnoreFailure,
+				},
+				CACerts: sat.CACerts,
+				Command: sat.Command,
+			}
+			if satDone {
+				return nil, fmt.Errorf("%w: satellite", ErrFirstbootAlreadySet)
+			}
+			fo.Scripts = append(fo.Scripts, opts)
+			satDone = true
+		}
+
+		if aap != nil {
+			opts := AAPFirstbootOptions{
+				FirstbootCommonOptions: FirstbootCommonOptions{
+					Name:          aap.Name,
+					IgnoreFailure: aap.IgnoreFailure,
+				},
+				CACerts:        aap.CACerts,
+				JobTemplateURL: aap.JobTemplateURL,
+				HostConfigKey:  aap.HostConfigKey,
+			}
+			if aapDone {
+				return nil, fmt.Errorf("%w: aap", ErrFirstbootAlreadySet)
+			}
+			fo.Scripts = append(fo.Scripts, opts)
+			aapDone = true
+		}
+	}
+
+	return fo, nil
+}
+
+func (fo *FirstbootOptions) Each(cbCustom func(CustomFirstbootOptions) error, cbSat func(SatelliteFirstbootOptions) error, cbAAP func(AAPFirstbootOptions) error) error {
+	for _, script := range fo.Scripts {
+		switch s := script.(type) {
+		case CustomFirstbootOptions:
+			if err := cbCustom(s); err != nil {
+				return err
+			}
+		case SatelliteFirstbootOptions:
+			if err := cbSat(s); err != nil {
+				return err
+			}
+		case AAPFirstbootOptions:
+			if err := cbAAP(s); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown firstboot option type: %T", script)
+		}
+	}
```

Comment:
I'm not sure about this idiom of having an interface to put everything into a single array then taking them out again. It feels a bit that you don't really gain anything since you only use `.Each` to access this? I'll have to think more when I'm more awake.

Reply:
I do not like it too, but keep in mind the BP allows you to order all firstboot scripts, including AAP and Satellite, so there has to be a common structure holding it. Unless I want to store indices and implement some kind of custom iteration, that sounds uglier than this.

What is a bit weird is the `Each` iterator, I have to agree. I am not sure about it, removing it will not get rid of the type assertion tho.